### PR TITLE
Kamino - Add incremental solve option to FK solver

### PR DESCRIPTION
**Pull request by Guirec-Maloisel**
## Description

To increase the robustness of the FK solver, in particular resetting to poses that are far from the rest pose, this introduces the option to decompose large jumps in actuator coordinates into several intermediary steps, by updating the target actuator coordinates incrementally at every Newton iteration.

This PR also includes two related bug fixes (that prevented modified unit tests from passing):
- The kernel resetting from base pose (within FK, as initial guess) was not fully general (was assuming zero/identity initial actuator coordinates) which led up to the base offset being applied twice in the case of a base body (not of a base joint).
- The computation of joint relative velocity in kinematics/joints.py (for velocity-level residuals) was missing a term, so velocity-level residuals that should have been zero were not.

Resolves vastsoun/newton#319

## Checklist

- [x] New or existing tests cover these changes
- [x] The documentation is up to date with these changes
- [x] `CHANGELOG.md` has been updated (if user-facing change)

## Test plan

This was tested on animatronic examples and was seen to improve the robustness significantly. The unit tests have also been altered to largely increase the sampling range of random poses, to demonstrate the added robustness.

## Summary by CodeRabbit

* **New Features**
  * Incremental actuator-coordinate solving with configurable linear/angular step limits and per-world minimum-iteration enforcement.
  * Support for incremental target actuator updates during solves.

* **Improvements**
  * Optional quaternion normalization and improved actuator-coordinate evaluation/correction (angle wrapping and quaternion sign canonicalization).
  * Adjusted body angular-velocity handling and stricter Newton convergence gating.

* **Tests**
  * FK tests updated with standardized actuator-coordinate canonicalization and tighter validation tolerances.

### CODE REVIEW

File: newton/_src/solvers/kamino/config.py [hunk 1]
```diff
@@ -837,6 +837,30 @@ class ForwardKinematicsSolverConfig:
...
 
+    use_incremental_solve: bool = True
+    """
+    Whether to automatically split large steps in actuator coordinates into smaller steps
+    in the FK solve, to improve the solver's robustness for a mild added cost.
+    Changes to this setting after the solver's initialization lead to undefined behavior.
+    Defaults to `True`.
+    """
+
+    max_linear_incremental_step_meters: float = 0.05
```
Comment (chschuma-disney) [inline newton/_src/solvers/kamino/config.py]: I would avoid the unit in the name of the config parameter, and document it in the docstring.

File: newton/_src/solvers/kamino/config.py [hunk 2]
```diff
@@ -837,6 +837,30 @@ class ForwardKinematicsSolverConfig:
...
+    """
+    If incremental solve is enabled, maximal allowed step in linear actuator coordinates
+    per solver iteration, in meters. A lower value results in more incremental steps.
+    Changes to this setting after the solver's initialization will have no effect.
+    Defaults to `0.1`.
+    """
+
+    max_angular_incremental_step_degrees: float = 10.0
```
Comment (chschuma-disney) [inline newton/_src/solvers/kamino/config.py]: The same here with the unit in the parameter name. Also, I believe Newton generally uses radians. I'm not sure we should break with that convention here.

File: newton/_src/solvers/kamino/_src/solvers/fk/kernels.py [hunk 3]
```diff
@@ -213,6 +234,331 @@ def _eval_fk_actuated_dofs_or_coords(
...
+    # Read data
+    coord_id = world_actuated_coord_offsets[wd_id] + coord_id_loc
+    q_prev = actuators_q_prev[coord_id]
+    q_next = actuators_q_next[coord_id]
+    delta = delta_q_max[coord_id]
+    it = iteration[wd_id]
+
+    # Interpolate coordinate
+    sign = wp.where(q_prev > q_next, -1.0, 1.0)
+    actuators_q_curr[coord_id] = sign * wp.min(sign * q_prev + wp.float32(it + 1) * delta, sign * q_next)
```
[the remark this answers is not in the document]
Comment (Guirec-Maloisel) [reply newton/_src/solvers/kamino/_src/solvers/fk/kernels.py:495]: Good point, even though this had no impact currently (the garbage values written were not read by the kernel evaluating position-control transforms) it's still cleaner to add an early return. I fixed as suggested.

File: newton/_src/solvers/kamino/_src/solvers/fk/kernels.py [hunk 4]
```diff
@@ -213,6 +234,331 @@ def _eval_fk_actuated_dofs_or_coords(
...
+
+
+@wp.func
+def min_iteration_op(v0: wp.float32, v1: wp.float32, d: wp.float32) -> wp.int32:
+    """Function returning the floor of the absolute division of (v1 - v0) by d."""
+    eps = 1e-7  # Epsilon to ensure that we round down if |v1 - v0| = d
+    return wp.int32(wp.floor((wp.abs(v1 - v0) - eps) / d))
```
[the remark this answers is not in the document]
Comment (Guirec-Maloisel) [reply newton/_src/solvers/kamino/_src/solvers/fk/kernels.py:502]: I don't want to handle this case because it only happens if the deltas set in the config by the user are invalid (<= 0). Instead I made sure that config.validate() is called within finalize() (in case some settings in the config were edited manually after the post_init but before passing it to the solver). It is also clearly documented that the settings in the config should not be altered (for most of them) after passing them to the solver. So I'm then free to assume without checks that these are valid within the solver code.

File: newton/_src/solvers/kamino/_src/solvers/fk/solver.py [hunk 5]
```diff
@@ -366,6 +370,16 @@ def eval_joint_pos_init(jt_id_prev):
...
+        world_num_actuated_coords = np.array(
+            [
+                joints_num_actuated_coords[first_joint_id[wd_id] : first_joint_id[wd_id + 1]].sum()
+                for wd_id in range(self.num_worlds)
+            ]
+        )
+        world_actuated_coord_offsets = np.concatenate(([0], world_num_actuated_coords.cumsum()))
+        self.num_actuated_coords_max = np.max(world_num_actuated_coords)
+        if self.num_actuated_coords_max == 0:  # No actuated coordinates
+            self.config.use_incremental_solve = 0  # Turn off incremental solve (corner case in unit tests)
```
[the remark this answers is not in the document]
Comment (Guirec-Maloisel) [reply newton/_src/solvers/kamino/_src/solvers/fk/solver.py]: I fixed this differently, by removing this corner case handling altogether, and instead making the tile size / number of tiles / block size computation more robust to also handle the size zero case (which was the reason this was introduced in the first place, as this lead to taking the log of zero otherwise)

File: newton/_src/solvers/kamino/config.py [hunk 6]
```diff
@@ -837,6 +837,30 @@ class ForwardKinematicsSolverConfig:
...
+    Defaults to `True`.
+    """
+
+    max_linear_incremental_step_meters: float = 0.05
+    """
+    If incremental solve is enabled, maximal allowed step in linear actuator coordinates
+    per solver iteration, in meters. A lower value results in more incremental steps.
+    Changes to this setting after the solver's initialization will have no effect.
+    Defaults to `0.1`.
+    """
```
[the remark this answers is not in the document]
Comment (Guirec-Maloisel) [reply newton/_src/solvers/kamino/config.py]: Good catch, I pushed a fix

File: newton/_src/solvers/kamino/config.py [hunk 7]
```diff
@@ -889,6 +913,10 @@ def validate(self) -> None:
             raise ValueError("`max_line_search_iterations` must be positive.")
         if self.tolerance <= 0.0:
             raise ValueError("`tolerance` must be positive.")
+        if self.max_linear_incremental_step_meters <= 0.0:
+            raise ValueError("`max_linear_incremental_step_meters` must be positive.")
+        if self.max_angular_incremental_step_degrees <= 0.0:
+            raise ValueError("`max_angular_incremental_step_degrees` must be positive.")
```
[the remark this answers is not in the document]
Comment (Guirec-Maloisel) [reply newton/_src/solvers/kamino/config.py]: Since the default values are not raising errors, these errors can only occur if the user explicitly set invalid values there, so I think it's still fair to raise an exception there even if these settings are superfluous based on other setting values.

**Review by chschuma-disney — COMMENTED** [review]
Looks good to me overall, great work! I just have two comments regarding the config parameters.

File: newton/_src/solvers/kamino/config.py (hunk 1 above)
Comment (vastsoun) [reply newton/_src/solvers/kamino/config.py]: Agreed.

**Review by vastsoun — DISMISSED** [review]
Great work @Guirec-Maloisel. I agree with @chschuma-disney regarding the naming of the variables though.

File: newton/_src/solvers/kamino/config.py (hunk 2 above)
Comment (Guirec-Maloisel) [reply newton/_src/solvers/kamino/config.py]: I ended up using degrees because I thought the default parameter would be easier to set and document then. I pushed a fix where I use math.radians(10.0), that's probably still readable then.

File: newton/_src/solvers/kamino/config.py (hunk 1 above)
Comment (Guirec-Maloisel) [reply newton/_src/solvers/kamino/config.py]: I removed it from the parameter (it was already also in the docstring)

**Review by chschuma-disney — APPROVED** [review]
LGTM

File: newton/_src/solvers/kamino/_src/solvers/fk/kernels.py [hunk 8]
```diff
@@ -213,6 +234,333 @@ def _eval_fk_actuated_dofs_or_coords(
...
+        wp.static(_make_typed_joint_transform_to_coords_func(JointDoFType.FREE))(pos_rel, q_rel, offset, output)
+    elif dof_type == FKJointDoFType.PRISMATIC:
+        wp.static(_make_typed_joint_transform_to_coords_func(JointDoFType.PRISMATIC))(pos_rel, q_rel, offset, output)
+    elif dof_type == FKJointDoFType.REVOLUTE:
+        wp.static(_make_typed_joint_transform_to_coords_func(JointDoFType.REVOLUTE))(pos_rel, q_rel, offset, output)
+    elif dof_type == FKJointDoFType.SPHERICAL:
+        wp.static(_make_typed_joint_transform_to_coords_func(JointDoFType.SPHERICAL))(pos_rel, q_rel, offset, output)
+    elif dof_type == FKJointDoFType.UNIVERSAL:
+        wp.static(_make_typed_joint_transform_to_coords_func(JointDoFType.UNIVERSAL))(pos_rel, q_rel, offset, output)
+
```
[the remark this answers is not in the document]
Comment (Guirec-Maloisel) [reply newton/_src/solvers/kamino/_src/solvers/fk/kernels.py:278]: No, this is only called below for actuator coordinates, and is not used outside of this file (a more general version already exists in kinematics/joints.py). Technically GIMBAL actuactors could exist, but are not supported in the FK solver, and the GIMBAL joint will eventually be removed soon.

**Review by vastsoun — APPROVED** [review]
LGTM!
